mods/taskbar-empty-space-clicks.wh.cpp

```diff
@@ -1936,55 +2064,112 @@ void StartProcess(const std::wstring &command)
+    const MouseClick &currentClick = g_mouseClickQueue[-1];
+
+    if (previousClick.type != MouseClick::Type::MOUSE || currentClick.type != MouseClick::Type::MOUSE)
     {
         return false;
     }
 
-    // Note: The reason why UIAutomation interface is used is that it reliably returns a className of the element clicked.
-    // If standard Windows API is used, the className returned is always Shell_TrayWnd which is a parrent window wrapping the taskbar.
-    // From that we can't really tell reliably whether user clicked on the taskbar empty space or on some UI element on that taskbar, like
-    // opened window, icon, start menu, etc.
+    // Check if the current event is within the double-click time and distance
+    bool result = abs(previousClick.position.x - currentClick.position.x) <= GetSystemMetrics(SM_CXDOUBLECLK) &&
+                  abs(previousClick.position.y - currentClick.position.y) <= GetSystemMetrics(SM_CYDOUBLECLK) &&
+                  ((currentClick.timestamp - previousClick.timestamp) <= GetDoubleClickTime());
+    return result;
+}
 
-    com_ptr<IUIAutomationElement> pWindowElement = NULL;
-    if (FAILED(g_pUIAutomation->ElementFromPoint(pointerLocation, pWindowElement.put())) || !pWindowElement)
+bool IsDoubleTap(const MouseClick &previousClick, const MouseClick &currentClick)
+{
+    LOG_TRACE();
+
+    if (previousClick.type != MouseClick::Type::TOUCH || currentClick.type != MouseClick::Type::TOUCH)
     {
-        LOG_ERROR(L"Failed to retrieve UI element from mouse click");
         return false;
     }
 
-    bstr_ptr className;
-    if (FAILED(pWindowElement->get_CurrentClassName(className.GetAddress())) || !className)
+    if (previousClick.hWnd != currentClick.hWnd)
     {
-        LOG_ERROR(L"Failed to retrieve the Name of the UI element clicked.");
         return false;
     }
-    LOG_DEBUG(L"Clicked UI element ClassName: %s", className.GetBSTR());
-    const bool taskbarClicked = (wcscmp(className.GetBSTR(), L"Shell_TrayWnd") == 0) ||                        // Windows 10 primary taskbar
-                                (wcscmp(className.GetBSTR(), L"Shell_SecondaryTrayWnd") == 0) ||               // Windows 10 secondary taskbar
-                                (wcscmp(className.GetBSTR(), L"Taskbar.TaskbarFrameAutomationPeer") == 0) ||   // Windows 11 taskbar
-                                (wcscmp(className.GetBSTR(), L"Windows.UI.Input.InputSite.WindowClass") == 0); // Windows 11 21H2 taskbar
-    if (!taskbarClicked)
+
+    HWND hWnd = currentClick.hWnd;
+    UINT dpi = GetDpiForWindow(hWnd);
+    float dpiScale = static_cast<float>(dpi) / 96.0f; // 96 DPI is the standard scaling factor
+
+    // GetSystemMetrics(SM_CXDOUBLECLK) is suitable just for mouse, not really for touch
+    const int MAX_POS_OFFSET_PX = 15;
+    // if user has hi-res screen, every slight movement result in big pixel offset
+    const int MAX_POS_OFFSET_PX_SCALED = dpiScale * MAX_POS_OFFSET_PX;
```

Comment:
Instead of using floating points, `MulDiv` is usually used for scaling, something like:
```cpp
    const int MAX_POS_OFFSET_PX_SCALED = MulDiv(MAX_POS_OFFSET_PX, dpi, 96);
```
It also takes care of rounding.